File: frontend/src/types/knowledge.ts
/** Knowledge base domain types for the BTagent frontend. */

/** Source types for knowledge documents */
export type KnowledgeSourceType =
  | "investigation_report"
  | "runbook"
  | "threat_profile"
  | "agency_profile"
  | "enrichment_data"
  | "playbook_log"
  | "conversation";

/** A document stored in the knowledge base */
export interface KnowledgeDocument {
  id: string;
  title: string;
  source_type: KnowledgeSourceType;
  token_count?: number;
  metadata?: Record<string, unknown>;
  created_at?: string | null;
  updated_at?: string | null;
}

/** Document with content and chunk count */
export interface KnowledgeDocumentDetail extends KnowledgeDocument {
  content: string;
  chunk_count: number;
}

/** A single search result from hybrid search */
export interface KnowledgeSearchResult {
  chunk_content: string;
  document_title: string;
  source_type: string;
  relevance_score: number;
  metadata: Record<string, unknown>;
  document_id: string;
  chunk_id: string;
}

/** Query response from the knowledge API */
export interface KnowledgeQueryResponse {
  query: string;
  results: KnowledgeSearchResult[];
  total_results: number;
}

/** Paginated document list response */
export interface KnowledgeDocumentListResponse {
  items: KnowledgeDocument[];
  total: number;
  page: number;
  page_size: number;
}

/** Ingest request body */
export interface KnowledgeIngestRequest {
  title: string;
  content: string;
  source_type: string;
  metadata?: Record<string, unknown>;
}

/** Ingest response */
export interface KnowledgeIngestResponse {
  id: string;
  title: string;
  source_type: string;
  token_count: number;
  message: string;
}

/** Source type display configuration */
export const SOURCE_TYPE_CONFIG: Record<
  KnowledgeSourceType,
  { label: string; color: string }
> = {
  investigation_report: { label: "Investigation Report", color: "blue" },
  runbook: { label: "Runbook", color: "green" },
  threat_profile: { label: "Threat Profile", color: "red" },
  agency_profile: { label: "Agency Profile", color: "purple" },
  enrichment_data: { label: "Enrichment Data", color: "yellow" },
  playbook_log: { label: "Playbook Log", color: "orange" },
  conversation: { label: "Conversation", color: "slate" },
};

File: shared/btagent_shared/types/knowledge.py
"""Knowledge base domain models for BTagent RAG system."""

from datetime import datetime
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field


class KnowledgeSourceType(StrEnum):
    """Types of documents that can be stored in the knowledge base."""

    INVESTIGATION_REPORT = "investigation_report"
    RUNBOOK = "runbook"
    THREAT_PROFILE = "threat_profile"
    AGENCY_PROFILE = "agency_profile"
    ENRICHMENT_DATA = "enrichment_data"
    PLAYBOOK_LOG = "playbook_log"
    CONVERSATION = "conversation"


class KnowledgeDocument(BaseModel):
    """A document stored in the knowledge base."""

    id: str
    title: str
    source_type: KnowledgeSourceType
    content: str
    metadata: dict[str, Any] = Field(default_factory=dict)
    token_count: int = 0
    created_at: datetime | None = None
    updated_at: datetime | None = None


class KnowledgeChunk(BaseModel):
    """A vector-embedded chunk of a knowledge document."""

    id: str
    document_id: str
    chunk_index: int
    content: str
    metadata: dict[str, Any] = Field(default_factory=dict)
    token_count: int = 0


class SearchResult(BaseModel):
    """A single result from hybrid search."""

    chunk_content: str
    document_title: str
    source_type: str
    relevance_score: float
    metadata: dict[str, Any] = Field(default_factory=dict)
    document_id: str = ""
    chunk_id: str = ""


class IngestRequest(BaseModel):
    """Request to ingest a document into the knowledge base."""

    title: str
    content: str
    source_type: str
    metadata: dict[str, Any] = Field(default_factory=dict)


class QueryRequest(BaseModel):
    """Request to query the knowledge base."""

    query: str
    top_k: int = Field(default=5, ge=1, le=50)
    source_type_filter: str | None = None


class QueryResponse(BaseModel):
    """Response from a knowledge base query."""

    query: str
    results: list[SearchResult]
    total_results: int
    answer: str | None = None
    citations: list[dict[str, Any]] = Field(default_factory=list)
